pkg/config/expand.go

```diff
@@ -162,57 +162,79 @@ func getModuleInputMap(inputs []modulereader.VarInfo) map[string]string {
+// a list, in which case output values are appended and flattened using HCL.
+// mod: "using" module as defined above
+// useMod: "used" module as defined above
+// useModGroupID: deployment group ID to which useMod belongs
+// modInputs: input variables as defined by the using module code
+// useOutputs: output values as defined by the used module code
+// settingsToIgnore: a list of module settings not to modify for any reason;
+//
+//	typical usage will be to leave explicit blueprint settings unmodified
 func useModule(
 	mod *Module,
 	useMod Module,
 	useModGroupID string,
-	modInputs map[string]string,
+	modInputs []modulereader.VarInfo,
 	useOutputs []modulereader.VarInfo,
-	settingsInBlueprint []string,
-) (usedVars []string) {
-	usedVars = []string{}
+	settingsToIgnore []string,
+) ([]string, error) {
+	usedVars := []string{}
+	modInputsMap := getModuleInputMap(modInputs)
 	for _, useOutput := range useOutputs {
 		settingName := useOutput.Name
 
-		// Skip settings explicitly defined by users
-		if slices.Contains(settingsInBlueprint, settingName) {
+		// Explicitly ignore these settings (typically those in blueprint)
+		if slices.Contains(settingsToIgnore, settingName) {
 			continue
 		}
 
-		// Skip settings that do not match module inputs
-		inputType, ok := modInputs[settingName]
+		// Skip settings that do not have matching module inputs
+		inputType, ok := modInputsMap[settingName]
 		if !ok {
 			continue
 		}
 
-		// if the input exists in the module doing the using, then we modify it
-		// according to these rules:
-		// if the setting has a value, then it was set by "use" because we have
-		// skipped anything explicitly in the blueprint. If the setting is a
-		// list of any type, then we want to concatenate these values in order
-		// they were used; otherwise, the setting should remain set to the
-		// existing used value
 		_, setByUse := mod.Settings[settingName]
 		modVarName := getModuleVarName(useModGroupID, useMod.ID, settingName)
 		isInputList := strings.HasPrefix(inputType, "list")
+
+		if !setByUse && !isInputList {
+			mod.Settings[settingName] = modVarName
+			usedVars = append(usedVars, settingName)
+		}
+
 		if isInputList {
-			if !setByUse {
-				// Input is a list, create an outer list for it
-				mod.Settings[settingName] = []interface{}{}
-				mod.createWrapSettingsWith()
-				mod.WrapSettingsWith[settingName] = []string{"flatten(", ")"}
+			if err := mod.addListValue(settingName, modVarName); err != nil {
+				return nil, err
 			}
-			// Append value list to the outer list
-			mod.Settings[settingName] = append(
-				mod.Settings[settingName].([]interface{}), modVarName)
-			usedVars = append(usedVars, settingName)
-		} else if !setByUse {
-			// If input is not a list, set value if not already set and continue
-			mod.Settings[settingName] = modVarName
 			usedVars = append(usedVars, settingName)
 		}
 	}
-	return
+	return usedVars, nil
```

Comment:
I strongly prefer to actually return the values as you are doing here. I think some readability is lost by removing the name from the return value in the signature. I would suggest we reclaim that readability by adding a line in the docstring for the return value:

```
returns: a list of variable names that were used during the process of calling this function
```

Reply:
I agree it ought to be in the docstrings and will add it. I think the internal style conventions for named parameters could be reasonably summarized as "there are cases where it is very clearly bad to use them, and there are cases where it is not-too-objectionable to use them."